File: Hough/linetest.m
clc 
clear all
RGB=imread('linea1.jpg');
[ri,g,b]=imsplit(RGB);
step=1;
ri=ri(1:step:end,1:step:end);

figure(1)
imshow(ri)

T=graythresh(ri);
bw=~im2bw(ri,T);

% figure(2)
% imshow(bw)

[M N]=size(bw);
rhomax=ceil((N^2+M^2)^0.5);
angulo=90
theta=[-angulo:step:angulo];
thetamax=length(theta);
H=zeros(thetamax,2*rhomax);
rho=[-rhomax:rhomax];

% figure(3)
% imshow(H)

for i=1:M
    for j=1:N
        if bw(i,j)==1
            for t=theta
                r=(ceil(j*cosd(t)+i*sind(t)))+1;
                H(t+angulo+1,r+rhomax+1)=H(t+angulo+1,r+rhomax+1)+1;
            end
        end
    end
end

figure(4)
imshow(imadjust(H/max(H(:)))','XData',theta,'YData',rho,'InitialMagnification','fit');
title('Hough transform');
xlabel('\theta'), ylabel('\rho');
axis on, axis normal, hold on;
colormap(gca,hot);
[H, T, R]=hough(bw,'RhoResolution',1,'Theta',-90:1:89);
figure(5)
imshow(imadjust(rescale(H)),'XData',T,'YData',R,'InitialMagnification','fit');
title('Hough transform');
xlabel('\theta'), ylabel('\rho');
axis on, axis normal, hold on;
colormap(gca,hot);

[a b]=find(H==max(H(:)))

rhoL=rho(a)
thetaL=theta(b)

x=[1:M]
y=-cotd(thetaL)*x+rhoL/sind(thetaL)

figure(6)
imshow(ri)
hold on
plot(x,y,'r--')
hold off

File: Hough/linetest.py
from ipFunctions import *
import matplotlib.pyplot as plt
import numpy as np

RGB = plt.imread('linea1.jpg')
ri, g, b = imsplit(RGB)
step = 1
ri = imundersize(ri,step)

plt.figure()
plt.set_cmap('gray')
plt.imshow(ri)

T=graythresh(ri)
bw=~im2bw(ri,T)

# figure(2)
# imshow(bw)

M, N = bw.shape
rhomax = np.ceil((N**2+M**2)**0.5)
angulo = 90
theta = np.arange(-angulo,angulo+1,step)
thetamax = len(theta)
H = np.zeros([int(thetamax),int(2*rhomax)])
rho = np.arange(-rhomax,rhomax+1)

# figure(3)
# imshow(H)

for i in range(1,N):
    for j in range(1,M):
        if bw[i,j]==1:
            for t in theta:
                r=(np.ceil(j*np.cos(np.radians(t))+i*np.sin(np.radians(t))))+1
                H[int(t+angulo),int(r+rhomax+1)]=H[int(t+angulo),int(r+rhomax+1)]+1

H = np.transpose(H)
a = imadjust(np.uint8(H / np.max(H)))
print(a)

plt.figure()
# plt.imshow(a,'XData',theta,'YData',rho,'InitialMagnification','fit')
plt.imshow(histeq(np.uint8(a/np.max(a))))
plt.title('Hough transform');
plt.xlabel('theta') 
plt.ylabel('rho')
# axis on, axis normal, hold on;
plt.set_cmap('hot')

a, b = np.where(H==np.max(H))
print(a,b)

rhoL=rho[a]
thetaL=theta[b]

x = np.arange(1,M)
y = - (1/(np.tan(np.radians(thetaL))))*x+rhoL/np.sin(np.radians(thetaL))

plt.figure()
plt.imshow(ri)
plt.plot(x,y,'r--')

plt.show()
